---
title: "idep Workflow"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{idep Workflow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r eval=TRUE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  echo = TRUE, #Change to TRUE to see code
  fig.height=5.5, 
  fig.width=7.5
)
```

```{r setup}
library(idepGolem)
```


```{r}
#devtools::load_all()
```

# LOAD DATA FUNCTIONS

```{r Create Data Paths}
idep_data <- get_idep_data()

#DATABASE <- Sys.getenv("IDEP_DATABASE")[1]
DATABASE <- "F:/IDEP_DATABASE/"
YOUR_DATA_PATH <- paste0(DATABASE, "data_go/BcellGSE71176_p53.csv")
YOUR_EXPERIMENT_PATH <- paste0(DATABASE, "data_go/BcellGSE71176_p53_sampleInfo.csv")

expression_file <- data.frame(
  datapath = YOUR_DATA_PATH
)
experiment_file <- data.frame(
  datapath = YOUR_EXPERIMENT_PATH
)

load_data <- input_data(
  expression_file = expression_file,
  experiment_file = experiment_file,
  go_button = FALSE,
  demo_data_file = idep_data$demo_data_file,
  demo_metadata_file = idep_data$demo_metadata_file
)

converted <- convert_id(
  query = rownames(load_data$data),
  idep_data = idep_data,
  select_org = "BestMatch"
)

all_gene_info <- gene_info(
  converted = converted,
  select_org = "BestMatch",
  idep_data = idep_data
)

converted_data <- convert_data(
  converted = converted,
  no_id_conversion = FALSE,
  data = load_data$data
)

gene_names <- get_all_gene_names(
  mapped_ids = converted_data$mapped_ids,
  all_gene_info = all_gene_info
)
```

## PRE-PROCESS FUNCTIONS

```{r Pre-Process Data}
processed_data <- pre_process(
  data = converted_data$data,
  missing_value = "geneMedian",
  data_file_format = 1,
  low_filter_fpkm = NULL,
  n_min_samples_fpkm = NULL,
  log_transform_fpkm = NULL,
  log_start_fpkm = NULL,
  min_counts = .5,
  n_min_samples_count = 1,
  counts_transform = 1,
  counts_log_start = 4,
  no_fdr = FALSE
)
```

```{r}
total_counts_ggplot(
  counts_data = processed_data$raw_counts,
  sample_info = load_data$sample_info
)
```

```{r fig.height=5.5, fig.width=7.5}
eda_scatter(
  processed_data = processed_data$data,
  #plot_xaxis = "p53_IR_1",
  #plot_yaxis = "p53_mock_1"
  plot_yaxis = colnames(processed_data$data)[1],
  plot_xaxis = colnames(processed_data$data)[2]
)
```

```{r fig.height=5.5, fig.width=7.5}
eda_boxplot(
  processed_data = processed_data$data,
  sample_info = load_data$sample_info
)
```

```{r fig.height=5.5, fig.width=7.5}
eda_density(
  processed_data = processed_data$data,
  sample_info = load_data$sample_info
)
```

```{r}
merged_raw_counts <- merge_data(
  all_gene_names = gene_names,
  data = processed_data$raw_counts,
  merge_ID = "ensembl_ID"
)

merged_data <- merge_data(
  all_gene_names = gene_names,
  data = processed_data$data,
  merge_ID = "ensembl_ID"
)
```

## PCA plots
```{r}
PCA_plot(
  data = processed_data$data,
  sample_info = load_data$sample_info,
  selected_color = "Sample_Name",
  selected_shape = "Sample_Name"
)
t_SNE_plot(
  data= processed_data$data,
  sample_info = load_data$sample_info,
  selected_color = "Sample_Name",
  selected_shape = "Sample_Name"
  )
pc_factor_correlation(
  data = processed_data$data,
  sample_info = load_data$sample_info
)
MDS_plot(
  data = processed_data$data,
  sample_info = load_data$sample_info,
  selected_color = "Sample_Name",
  selected_shape = "Sample_Name"
)
```

# PCAtools plots
```{r, eval=FALSE}
#devtools::load_all()
PCA_biplot(
  data = processed_data$data,
  sample_info = load_data$sample_info,
  select_gene_id = "symbol",
  all_gene_names = gene_names
  )
PCA_Scree(processed_data = processed_data$data)
# 
PCAtools_eigencorplot(
  processed_data = processed_data$data,
  sample_info = load_data$sample_info
      )
```


#### SEARCHING
```{r}
heatmap_data <- process_heatmap_data(
  data = processed_data$data,
  n_genes_max = 50,
  gene_centering = TRUE,
  gene_normalize = FALSE,
  sample_centering = FALSE,
  sample_normalize = FALSE,
  all_gene_names = gene_names,
  select_gene_id = "ensembl_ID"
)
dist_funs <- dist_functions()
hclust_funs <- hcluster_functions()

gene_sets <- read_pathway_sets(
  all_gene_names_query = gene_names[1:500, ],
  converted = converted,
  go = "GOBP",
  select_org = "BestMatch",
  gmt_file = NULL,
  idep_data = idep_data,
  gene_info = all_gene_info
)

pathway_info <- find_overlap(
  pathway_table = gene_sets$pathway_table,
  query_set = gene_sets$query_set,
  total_genes = gene_sets$total_genes,
  processed_data = processed_data$data,
  gene_info = all_gene_info,
  go = "GOBP",
  idep_data = idep_data,
  sub_pathway_files = gene_sets$pathway_files,
  use_filtered_background = TRUE,
  select_org = "BestMatch",
  reduced = .75
)

#pathway_info
```


```{r}
limma <- limma_value(
  data_file_format = "1",
  counts_deg_method = "3",
  raw_counts = processed_data$raw_counts,
  limma_p_val = .1,
  limma_fc = 2,
  select_model_comprions = "p53: NULL vs. WT",
  sample_info = load_data$sample_info,
  select_factors_model = c("p53", "Treatment"),
  select_interactions = "p53:Treatment",
  select_block_factors_model = NULL,
  factor_reference_levels = c("p53:WT", "Treatment:MOCK"),
  processed_data = processed_data$data,
  counts_log_start = 4,
  p_vals = processed_data$p_vals
)

contrast_samples <- find_contrast_samples(
  select_contrast = "NULL-WT", 
	all_sample_names = colnames(processed_data$data),
	sample_info = load_data$sample_info,
	select_factors_model = c("p53", "Treatment"),
	select_model_comprions = "p53: NULL vs. WT", 
  reference_levels = c("p53:WT", "Treatment:MOCK"),
	counts_deg_method = "3",
	data_file_format = "1"
)

deg_heat <- deg_heat_data(
  limma = limma,
  select_contrast = "NULL-WT",
  processed_data = processed_data$data,
  contrast_samples = contrast_samples
)

up_reg_data <- pathway_info
up_reg_data <- as.data.frame(up_reg_data)
up_reg_data$direction <- rep("Up", nrow(up_reg_data))
#up_reg_data
down_reg_data <- deg_heat$genes[deg_heat$bar == -1, ]
down_reg_data <- as.data.frame(down_reg_data)
down_reg_data$direction <- rep("Down", nrow(down_reg_data))
#down_reg_data

# main_heat_deg <- deg_heatmap(
#   data = deg_heat$genes,
#   bar = deg_heat$bar,
#   heatmap_color_select = c("green", "black", "red")
# )
# ComplexHeatmap::draw(main_heat_deg)

plot_volcano(
  select_contrast = "NULL-WT",
  comparisons = limma$comparisons,
  top_genes = limma$top_genes,
  limma_p_val = .1,
  limma_fc = 2, 
  plot_colors = c("red", "grey", "blue")
)

plot_ma(
  select_contrast = "NULL-WT",
  comparisons = limma$comparisons,
  top_genes = limma$top_genes,
  limma_p_val = .1,
  limma_fc = 2,
  contrast_samples = contrast_samples,
  processed_data = processed_data$data, 
  plot_colors = c("red","yellow", "blue")
)
```

```{r}
comp_venn <- list_comp_venn(
  limma = limma, 
  up_down_regulated = TRUE
)


plot_venn(
  limma = limma,
  up_down_regulated = TRUE, 
  select_comparisons_venn = comp_venn$choices_first_three
)
```


```{r}
results <- limma$results
select_comparisons_venn = comp_venn$choices_first_three

result_up <- results
result_up[result_up < 0] <- 0
colnames(result_up) <- paste0("Up_", colnames(result_up))
result_down <- results
result_down[result_down > 0] <- 0
colnames(result_down) <- paste0("Down_", colnames(result_down))
results <- cbind(result_up, result_down)

 ixa <- c()
  for (comps in select_comparisons_venn) {
    # If not interaction term
    if (!grepl("^I:|^I-|^Up_I:|^Up_I-|^Down_I:|^Down_I-", comps)) {
      ix <- match(comps, colnames(results))
    } else {
      # Mismatch in comparison names for interaction terms for DESeq2
      # I:water_Wet.genetic_Hy   in the selected Contrast
      # Diff-water_Wet-genetic_Hy  in column names
      tem <- gsub("^I-", "I:", colnames(results))
      tem <- gsub("-", "\\.", tem)
      ix <- match(comps, tem)

      # This is for limma package
      if (is.na(ix)) {
        ix <- match(comps, colnames(results))
      }
    }
    ixa <- c(ixa, ix)
  }
 
 
  # Only use selected comparisons
  results2 <- results[, ixa, drop = FALSE]
  if (dim(results)[2] > 5) {
    results <- results[, 1:5]
  }
  colnames(results) <- gsub("^I-", "I:", colnames(results))

```


```{r}

results <- limma$results


select_comparisons_venn = comp_venn$choices_first_three

result_up <- results
result_up[result_up < 0] <- 0
colnames(result_up) <- paste0("Up_", colnames(result_up))
result_down <- results
result_down[result_down > 0] <- 0
colnames(result_down) <- paste0("Down_", colnames(result_down))
results <- cbind(result_up, result_down)

temp <- results %>%
  as_tibble() %>%
  mutate(id = 1:n()) %>%
  gather(Group, GroupMember, `Up_NULL-WT`:`Down_I:p53_NULL-Treatment_IR`) %>% 
  group_by_at(vars(-c(Group, GroupMember))) %>%
  nest() %>%
  mutate(Group =  map(data, ~ if(all(.x$GroupMember == 0)){ character(0) }else{ .x$Group[.x$GroupMember == 1]} )) %>% 
  select(-data, -id)

ggplot(temp, aes(x = Group)) + 
  geom_bar(fill = "blue") + 
  geom_text(stat = "count", aes(label = after_stat(count)), vjust = -1, size = 3) + 
  scale_y_continuous(limits = c(0, 1350)) + 
  scale_x_upset(n_intersections = 20)
  

switch <- function(value){
  ifelse(value != 0, colnames(value), value)
}

colnam <- colnames(results)

test <- results %>% 
  as_tibble() %>% 
  mutate(across(everything(), .fns = as.character))

for (i in 1:length(colnam)) {
  test[test[,i] != 0, i] <- colnam[i]
  test[test[,i]  == 0, i] <- NA
}

x <- split(as.data.frame(test), 1:nrow(test))

my_fun <- function(data, colnam){
  list <- c() 
  for (i in length(colnam)) {
    list <- c(list, data[i])
  }
}

for (i in 1:dim(test)[1]){
   test[i, length(colnam) + 1] = my_fun(test[i, ], colnam)
}
```


```{r}
results <- limma$results
select_comparisons_venn = comp_venn$choices_first_three

result_up <- results
result_up[result_up < 0] <- 0
colnames(result_up) <- paste0("Up_", colnames(result_up))
result_down <- results
result_down[result_down > 0] <- 0
colnames(result_down) <- paste0("Down_", colnames(result_down))
results <- cbind(result_up, result_down)

options <- colnames(results)


test <- results %>%
  mutate(list = list)
```


```{r}
library(ggupset)
library(ggplot2)
library(tidyverse)

df <- tidy_movies %>%
  dplyr::distinct(title, year, length, .keep_all = TRUE)

ggplot(df, aes(x = Genres)) + 
  geom_bar(fill = "blue") + 
  geom_text(stat = "count", aes(label = after_stat(count)), vjust = -1, size = 3) + 
  scale_y_continuous(limits = c(0, 1350)) + 
  scale_x_upset(n_intersections = 20)
```



# Pathway Tab Stuff
```{r}
gene_sets <- read_gene_sets(
  converted = converted,
  all_gene_names = gene_names,
  go = "GOBP",
  select_org = "BestMatch",
  idep_data = idep_data,
  my_range = c(15, 2000)
)

gage <- gage_data(
  select_go = "GOBP",
  select_contrast = "I:p53_NULL.Treatment_IR",
  min_set_size = 15,
  max_set_size = 2000,
  limma = limma,
  gene_p_val_cutoff = 1,
  gene_sets = gene_sets,
  absolute_fold = FALSE,
  pathway_p_val_cutoff = .1,
  n_pathway_show = 30
)

contrast_samples_pgsea <- find_contrast_samples(
  select_contrast = "I:p53_NULL.Treatment_IR", 
	all_sample_names = colnames(processed_data$data),
	sample_info = load_data$sample_info,
	select_factors_model = c("p53", "Treatment"),
	select_model_comprions = "p53: NULL vs. WT", 
  reference_levels = c("p53:WT", "Treatment:MOCK"),
	counts_deg_method = "3",
	data_file_format = "1"
)

#plot_pgsea(
  #my_range = c(15, 2000),
  #processed_data = processed_data$data,
  #contrast_samples = contrast_samples_pgsea,
  #gene_sets = gene_sets,
  #pathway_p_val_cutoff = .1,
  #n_pathway_show = 30
#)


# fgsea <- fgsea_data(
#   select_contrast = "I:p53_NULL.Treatment_IR",
#   my_range = c(15, 2000),
#   limma = limma,
#   gene_p_val_cutoff = 1,
#   gene_sets = gene_sets,
#   absolute_fold = FALSE,
#   pathway_p_val_cutoff = .1,
#   n_pathway_show = 30
# )

# reactome_data(
#   select_contrast = "I:p53_NULL.Treatment_IR",
#   my_range = c(15, 2000),
#   limma = limma,
#   gene_p_val_cutoff = 1,
#   converted = converted,
#   idep_data = idep_data,
#   pathway_p_val_cutoff = .1,
#   n_pathway_show = 30,
#   absolute_fold = FALSE
# )

#pgsea_plot_all(
  #go = "GOBP",
  #my_range = c(15, 2000),
  #data = processed_data$data,
  #select_contrast = "I:p53_null.Treatment_IR",
  #gene_sets = gene_sets,
  #pathway_p_val_cutoff = .1,
  #n_pathway_show = 30
#)

# pgsea_plot_data <- get_pgsea_plot_data(
#   my_range = c(15, 2000),
#   data = processed_data$data,
#   select_contrast = "I:p53_NULL.Treatment_IR",
#   gene_sets = gene_sets,
#   sample_info = load_data$sample_info,
#   select_factors_model = c("p53", "Treatment"),
#   select_model_comprions = "p53: NULL vs. WT",
#   pathway_p_val_cutoff = .1,
#   n_pathway_show = 30
# ) 

gage[, 2]

select_data <- pathway_select_data(
  sig_pathways = "RNA processing ",
  gene_sets = gene_sets,
  contrast_samples = contrast_samples_pgsea,
  data = processed_data$data,
  select_org = "BestMatch",
  all_gene_names = gene_names
)

basic_heatmap(
  data = select_data,
  c("green", "black", "red")
)

pathway_list_data <- get_pathway_list_data(
  pathway_method = 1,
  gage_pathway_data = gage,
  fgsea_pathway_data = NULL,
  pgsea_plot_data = NULL,
  pgsea_plot_all_samples_data = NULL,
  go = "GOBP",
  select_org = "BestMatch",
  gene_info = all_gene_info,
  gene_sets = gene_sets
)

enrichment_plot(
  go_table = pathway_list_data
)

# kegg_pathway(
#   go = "KEGG",
#   gage_pathway_data = gage,
#   sig_pathways = "Oocyte meiosis",
#   select_contrast = "I:p53_NULL.Treatment_IR",
#   limma = limma,
#   converted = converted,
#   idep_data = idep_data,
#   select_org = "BestMatch"
# )
```

# Genome Tab
```{r, include=FALSE}
# all_gene_info <- get_gene_info(
#   converted = converted,
#   select_org = "BestMatch",
#   gene_info_files = idep_data$gene_info_files
# )
# 
# chromosome_plotly(
#   limma = limma,
#   select_contrast = "I:p53_NULL.Treatment_IR",
#   all_gene_info = all_gene_info,
#   ignore_non_coding = FALSE,
#   limma_p_val_viz = .1,
#   limma_fc_viz = 4,
#   label_gene_symbol = FALSE,
#   ma_window_size = 6,
#   ma_window_steps = 2,
#   ch_region_p_val = .01
# )
# 
# genome_plot_data_pre <- get_genome_plot_data_pre(
#   select_contrast = "I:p53_NULL.Treatment_IR",
#   limma = limma,
#   all_gene_info = all_gene_info
# )
# 
# genome_plot_data <- get_genome_plot_data(
#   genome_plot_data_pre = genome_plot_data_pre,
#   all_gene_info = all_gene_info,
#   select_contrast = "I:p53_NULL.Treatment_IR",
#   limma = limma,
#   regions_p_val_cutoff = .01,
#   statistic_cutoff = .5
# )
# 
# get_genome_plot(
#   genome_plot_data = genome_plot_data,
#   regions_p_val_cutoff = .01,
#   statistic_cutoff = .5
# )
```

# Bicluster Functions
```{r}
#"BCCC()" biclust
#"BCQU()" QUBIC
#"BCUnibic()" runibic
#"BCXmotifs()" biclust
#"BCPlaid()" biclust
#"BCSpectral()" biclust
#"BCBimax()" biclust
#"BCQuest()" biclust
biclustering <- get_biclustering(
  data = processed_data$data,
  n_genes = 1000,
  biclust_method = "biclust::BCCC()"
)

biclust_data <- biclust::bicluster(biclustering$data, biclustering$res, 1)[[1]]

basic_heatmap(
  data = biclust_data,
  c("green", "black", "red")
)

biclust_gene_names <- merge_data(
  all_gene_names = gene_names,
  data = biclust_data,
  merge_ID = "ensembl_ID"
)
# Only keep the gene names and scrap the data
gene_names_query <- dplyr::select_if(biclust_gene_names, is.character)

gene_sets <- read_pathway_sets(
  all_gene_names_query = gene_names_query,
  converted = converted,
  go = "GOBP",
  select_org = "BestMatch",
  gmt_file = NULL,
  idep_data = idep_data,
  gene_info = all_gene_info
)

pathway_info <- find_overlap(
  pathway_table = gene_sets$pathway_table,
  query_set = gene_sets$query_set,
  total_genes = gene_sets$total_genes,
  processed_data = processed_data$data,
  gene_info = all_gene_info,
  go = "GOBP",
  idep_data = idep_data,
  select_org = "GOBP",
  sub_pathway_files = gene_sets$pathway_files,
  use_filtered_background = FALSE,
  reduced = FALSE
)

# biclust_table <- get_biclust_table_data(
#   res = biclustering$res,
#   biclust_data = biclust_data,
#   select_go = "GOBP",
#   select_org = "BestMatch",
#   all_gene_info = all_gene_info
# )
```

# NETWORK TAB
```{r}
wgcna <- get_wgcna(
  data = processed_data$data,
  n_genes = 1000,
  soft_power = 5,
  min_module_size = 20
)

get_wgcna_modules(wgcna)

network_plot <- get_network_plot(
  select_wgcna_module = "1. turquoise (495 genes)",
  wgcna = wgcna,
  top_genes_network = 10,
  select_go = "GOBP",
  select_org = "BestMatch",
  all_gene_info = all_gene_info,
  edge_threshold = .4
)

network_plot()

network_query <- network_enrich_data(
  select_wgcna_module = "1. turquoise (495 genes)",
  wgcna = wgcna
)

gene_query <- dplyr::filter(gene_names, ensembl_ID %in% network_query)

gene_sets <- read_pathway_sets(
  all_gene_names_query = gene_query,
  converted = converted,
  go = "GOBP",
  select_org = "BestMatch",
  gmt_file = NULL,
  idep_data = idep_data,
  gene_info = all_gene_info
)

pathway_info <- find_overlap(
  pathway_table = gene_sets$pathway_table,
  query_set = gene_sets$query_set,
  total_genes = gene_sets$total_genes,
  processed_data = processed_data$data,
  gene_info = all_gene_info,
  go = "GOBP",
  idep_data = idep_data,
  select_org = "BestMatch",
  sub_pathway_files = gene_sets$pathway_files,
  use_filtered_background = TRUE,
  reduced = FALSE
)

plot_scale_independence(
  wgcna = wgcna
)

plot_mean_connectivity(
  wgcna = wgcna
)
```


		