---
title: "idep Workflow"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{idep Workflow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r eval=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
#library(idepGolem)
```


#### Introduction

The purpose of this project is to provide the tools to examine expression data
with simple and effective features. idep enables the user to upload their data
and go to work processing and analyzing the data to draw informative
conclusions. This workflow is visualized in the chart below. 

<center>

![](flowchart.png)

<\center>

## Load Data

This function will enable us to use the functions that drive the idepGolem 
package.
```{r}
devtools::load_all()
```
<br>

#### Data Paths

This begins the process of loading the data. `get_idep_data()` returns a list of
the data that will be used to process and examine the loaded expression data. To
analyze a data set, just substitute the path to the expression data into
`YOUR_DATA_PATH`. If your data comes with information about the samples,
substitute the path for that file into `YOUR_EXPERIMENT_PATH`. These data paths
are then stored in a data frame that will be used in the `input_data` function.
The structure of your data should be a **numeric** matrix with gene IDs as rows
and sample names as columns.
```{r Create Data Paths}
idep_data <- get_idep_data()

YOUR_DATA_PATH <- "E:/idep_9_24/data/data103/data_go/BcellGSE71176_p53.csv"
YOUR_EXPERIMENT_PATH <- "E:/idep_9_24/data/data103/data_go/BcellGSE71176_p53_sampleInfo.csv"

expression_file <- data.frame(
  datapath = YOUR_DATA_PATH
)
experiment_file <- data.frame(
  datapath = YOUR_EXPERIMENT_PATH
)
```
<br>

#### **`input_data`** Function

`input_data` takes in the established data paths and reads it into a data set to
analyze. If an experiment file is provided, the data will be returned in a list
containing the expression data and sample information. To run idep on an example
data set, set the value of `go_button` to `TRUE`. The `demo_data_file` and
`demo_metadata_file` are the built-in paths to the example data set and should
not be changed. The returned expression will be a numeric matrix with gene IDs
as rownames. The columns will denote the sample.
```{r Load Data}
loaded_data <- input_data(
  expression_file = expression_file,
  experiment_file = experiment_file,
  go_button = FALSE,
  demo_data_file = idep_data$demo_data_file,
  demo_metadata_file = idep_data$demo_metadata_file
)
```
<br>

#### Gene ID and Meta Data Functions

Idep works with "ensembl" gene IDs to perform many of its functions. Given a
data set, idep will convert IDs to ensembl and then gather metadata for each
gene if the database recognizes the species. This process occurs in the
following functions. To search for a species that fits the expression data,
utilize the line 
`idep_data$species_choice[grep("Mouse", names(idep_data$species_choice))]` and
input the numeric code as a character string (i.e 107 -> "107") in `select_org`.

1. **`convert_id`**

    This function will take in a 'query' and use the 'idep_data' to match the 
    provided gene IDs with the corresponding ensembl ID. The IDs from the
    inputted data are currently the rownames of the data frame `data` stored in
    the list `loaded_data`. `select_org` is a parameter to provide the species
    that the expression data is from. `convert_id` returns a list containing
    information about the query results. The converted IDs are are accessed by
    `converted$ids`. `print(converted$species_matched[1,])` shows the species
    that was recognized by the provided gene IDs.

2. **`gene_info`**

    Gathering the information on the converted ensembl IDs is the next step. 
    The `converted` results are passed into the `gene_info` function to retrieve
    gene name information that is stored in `idep_data` as well. If `select_org`
    is `BestMatch`, the function will use the matched species from converted to
    retrieve the gene data. If a species code is provided, it will use that
    species directly. `gene_info` will then read a gene data file from the
    `idep_data` and return the resulting data frame. This data will be used to
    map to more informative gene names in the coming functions.
    
3. **`convert_data`**

    This step will return the original data from `input_data` with the converted
    ensembl IDs as rownames, and a data frame of the mapped IDs. It uses the IDs
    that were matched with `convert_id` to swap the current rownames with the 
    corresponding ensembl ID. `no_id_conversion` can be set to `TRUE` if the 
    desired IDs are the original ones, but the simpler solution is to simply not
    run the function. The returned list contains the data with converted IDs in 
    `data` and IDs in `mapped_ids`.

4. **`get_all_gene_names`**

    This function enables us to combine all the gene name information that has
    been gathered in the preceding functions. `mapped_ids` are the matched IDs
    from `converted`, and `all_gene_info` is the data read from idep based off
    the identified species. This function returns a data frame with `User_ID`
    as the original IDs, `ensembl_ID` containing the converted IDs, and `symbol`
    containing the informative gene names. This data frame will later be used to
    create a data frame after processing for reproducibility.
    
```{r Data Manipulation}
# Species search for input data
idep_data$species_choice[grep("Mouse", names(idep_data$species_choice))]

converted <- convert_id(
  query = rownames(loaded_data$data),
  idep_data = idep_data,
  select_org = "BestMatch"
)

print(converted$species_matched[1,])

all_gene_info <- gene_info(
  converted = converted,
  select_org = "BestMatch",
  idep_data = idep_data
)

converted_data <- convert_data(
  converted = converted,
  no_id_conversion = FALSE,
  data = loaded_data$data
)

gene_names <- get_all_gene_names(
  mapped_ids = converted_data$mapped_ids,
  all_gene_info = all_gene_info
)
```
<br>

## Process Data

Now that the gene information has been retrieved and the IDs converted, it is
possible to process the data and perform a transformation for analysis.

<br>

#### **`pre_process`** Function

Processing the data requires many options to ensure the data gets to the desired
form. This is the data that will be used in the exploratory analysis to come.
The data to be processed is the `converted_data` from above. The parameters
`missing_value` and `data_file_format` have a list of input values to select 
from for processing. These values and descriptions can be found in the tables 
below. `missing_value` determines how missing data entries will be filled in, 
and `data_file_format` tells the function what type of data is being processed.

<div style="text-align: left">

Input (`missing_value`)    Description
-------------------------  -----------------------------------------------------------
`geneMedian`               Substitute median expression value for gene across all samples
`treatAsZero`              Substitute "0" for all missing values
`geneMedianInGroup`        Substitute median expression value for gene within sample group

Input (`data_file_format`)   Description
---------------------------  -----------------------------------------------------------------------
`1`                          Read counts data (recommended)
`2`                          Normalized expression values (RNA-seq FPKM, micro-array, etc.)
`3`                          Fold-changes and corrected P values from CuffDiff or any other program

</div>

<br>

##### **Read counts data**

For read counts data, the parameters `min_counts`, `n_min_samples_counts`,
`counts_transform` and `counts_log_start` need input values. `min_counts` and
`n_min_samples` work to filter the genes out of the data that do not have
sufficient counts. The function counts the columns that have more counts than
`min_counts` for each gene and filters out the genes with less columns than
`n_min_samples`. For example, if you only wanted genes with more than 0 counts
for each sample, you would set `min_counts = .5` and 
`n_min_samples = ncol(converted_data)`. `counts_transform` specifies the
transformation to perform on the counts data. This parameter has 3 input options
which are outlined below. If `counts_transform = 1`, `counts_log_start`
determines the constant to add to CPM to ensure that there are no `log(0)`
errors. All other parameters can be set to `NULL`

Input (`counts_transform`)   Description
---------------------------  --------------------------------------------------
`1`                          EdgeR: log2(CPM+c) (Counts per Million + constant)
`2`                          VST: variance stabilizing transform
`3`                          rlog: regularized log (slow)

<br>

##### **Normalized expression values Parameters**

If the data being processed is for normalized expression values, the parameters
`low_filter_fpkm`, `n_min_samples_fpkm`, `log_transform_fpkm` and 
`log_start_fpkm` need input values. `low_filter_fpkm` and `n_min_samples_fpkm`
work together to filter out the genes that do not have enough counts. The
function will count the columns that are greater than `low_filter_fpkm` and 
discard the gene if it is not more than `n_min_samples_fpkm` columns. The
parameter `log_transform_fpkm` is `TRUE/FALSE` for whether to perform a log
transform on the data or not. If `log_transform = TRUE`, then `log_start_fpkm`
is the value that will be added to ensure there are no `log(0)` errors. All
other values can be set to `NULL`.

<br>

##### **Fold-changes and corrected P values**

For this data type, there is only one additional parameter to provide an input
for, `no_fdr`. This is a logical `TRUE/FALSE` to specify whether the data
contains corrected p-values or not. `TRUE` denotes fold-change only data. All
other parameters can be set to `NULL`.

```{r Pre-Process Data}
processed_data <- pre_process(
  data = converted_data$data,
  missing_value = "geneMedian",
  data_file_format = 1,
  low_filter_fpkm = NULL,
  n_min_samples_fpkm = NULL,
  log_transform_fpkm = NULL,
  log_start_fpkm = NULL,
  min_counts = .5,
  n_min_samples_count = 1,
  counts_transform = 1,
  counts_log_start = 4,
  no_fdr = NULL
)
```
<br>

## Exploratory Data Analysis

Now that the data has been processed, it is possible to begin data analysis. It
is possible to create unique and informative plots with the processed data, and
idep has built in functions to assist with that.

<br>

#### **`total_counts_ggplot`** Function

This function will only work for read counts data, so if that is not the format
of the processed data, please skip this function. For read counts data, this
function creates a barplot to show the counts in millions for each sample in the
data. `counts_data` is the raw counts that is returned from the `pre_process`
function. `sample_info` is the experiment design data from the `input_data`
function. 

```{r fig.height=5.5, fig.width=7.5}
total_counts_ggplot(
  counts_data = processed_data$raw_counts,
  sample_info = loaded_data$sample_info
)
```

<br>

#### **`eda_scatter`** Function

Another useful plot is a scatter plot of the transformed expression for two
samples in the data. `eda_scatter` provides a scatter plot of two specified
sample (columns) of the processed data. `plot_xaxis` and `plot_yaxis` are
string names of the columns to be plotted. The options for these inputs can be
seen using `colnames(processed_data$data)`. `processed_data` is the returned
data from the `pre_process` function. 

```{r fig.height=5.5, fig.width=7.5}
colnames(processed_data$data)

eda_scatter(
  processed_data = processed_data$data,
  plot_xaxis = "p53_IR_1",
  plot_yaxis = "p53_mock_1"
)
```

<br>

#### **`eda_boxplot`** Function

`eda_boxplot` will show the distribution of the counts data after the 
pre-processing transformation. This function returns a ggplot boxplot of each
column in the data. `processed_data` is the transformed counts and `sample_info`
is the experiment design information from the `input_data` function. 

```{r fig.height=5.5, fig.width=7.5}
eda_boxplot(
  processed_data = processed_data$data,
  sample_info = loaded_data$sample_info
)
```

<br>

#### **`eda_density`** Function

This plotting function provides a plot of the distribution of counts values for
each sample in the data. The information is displayed in a density ridge plot in
the ggplot format. Again, the data provided is returned from the `pre_process`
function. The `sample_info` provides information on each sample in the dataset.

```{r fig.height=5.5, fig.width=7.5}
eda_density(
  processed_data = processed_data$data,
  sample_info = loaded_data$sample_info
)
```

<br>

#### `merge_data` Function

The current data is using the converted ensembl IDs as rownames in a large data
matrix, but the following function will create a data frame containing the gene
names from the `get_all_gene_names` function. `merge_data` takes the
`gene_names` data frame and a data matrix with ID rownames and creates a data
frame with columns for each ID that idep has matched from the original IDs.
This is useful to view different IDs for the same gene. It also enables writing
CSV files that can be saved and be used to reproduce results from idep. The
following code will merge the gene names with the raw counts data and the
transformed data from `pre_process`.

```{r}
merged_raw_counts <- merge_data(
  all_gene_names = gene_names,
  data = processed_data$raw_counts,
  merge_ID = "ensembl_ID"
)

merged_data <- merge_data(
  all_gene_names = gene_names,
  data = processed_data$data,
  merge_ID = "ensembl_ID"
)
```

#### SEARCHING
```{r}
# Gene search for individual plotting
merged_data[grep("H4c1", merged_data$symbol),]

heatmap_data <- process_heatmap_data(
  data = processed_data$data,
  n_genes_max = 50,
  n_genes_min = 0,
  gene_centering = TRUE,
  gene_normalize = FALSE,
  sample_centering = FALSE,
  sample_normalize = FALSE,
  all_gene_names = gene_names,
  select_gene_id = "ensembl_ID"
)
dist_funs <- dist_functions()
hclust_funs <- hcluster_functions()

devtools::load_all()

gene_sets <- read_pathway_sets(
  all_gene_names_query = gene_names[1:500, ],
  converted = converted,
  go = NULL,
  select_org = "BestMatch",
  gmt_file = NULL,
  idep_data = idep_data,
  gene_info = all_gene_info
)

pathway_info <- find_overlap(
  pathway_table = gene_sets$pathway_table,
  query_set = gene_sets$query_set,
  total_genes = gene_sets$total_genes,
  processed_data = processed_data$data,
  gene_info = all_gene_info,
  go = "GOBP",
  idep_data = idep_data,
  sub_pathway_files = gene_sets$pathway_files,
  use_filtered_background = TRUE,
  select_org = "BestMatch",
  reduced = .75
)
pathway_info
```


```{r}
limma <- limma_value(
  data_file_format = "1",
  counts_deg_method = "3",
  raw_counts = processed_data$raw_counts,
  limma_p_val = .1,
  limma_fc = 2,
  select_model_comprions = "p53: null vs. wt",
  sample_info = loaded_data$sample_info,
  select_factors_model = c("p53", "Treatment"),
  select_interactions = "p53:Treatment",
  select_block_factors_model = NULL,
  factor_reference_levels = c("p53:wt", "Treatment:mock"),
  processed_data = processed_data$data,
  counts_log_start = 4,
  p_vals = processed_data$p_vals
)

contrast_samples <- find_contrast_samples(
  select_contrast = "null-wt", 
	all_sample_names = colnames(processed_data$data),
	sample_info = loaded_data$sample_info,
	select_factors_model = c("p53", "Treatment"),
	select_model_comprions = "p53: null vs. wt", 
  reference_levels = c("p53:wt", "Treatment:mock"),
	counts_deg_method = "3",
	data_file_format = "1"
)

deg_heat <- deg_heat_data(
  limma = limma,
  select_contrast = "null-wt",
  processed_data = processed_data$data,
  contrast_samples = contrast_samples
)

up_reg_data <- pathway_info
up_reg_data <- as.data.frame(up_reg_data)
up_reg_data$direction <- rep("Up", nrow(up_reg_data))
up_reg_data
down_reg_data <- deg_heat$genes[deg_heat$bar == -1, ]
down_reg_data <- as.data.frame(down_reg_data)
down_reg_data$direction <- rep("Down", nrow(down_reg_data))
down_reg_data

main_heat_deg <- deg_heatmap(
  data = deg_heat$genes,
  bar = deg_heat$bar,
  heatmap_color_select = c("green", "black", "red")
)
ComplexHeatmap::draw(main_heat_deg)

plot_volcano(
  select_contrast = "null-wt",
  comparisons = limma$comparisons,
  top_genes = limma$top_genes,
  limma_p_val = .1,
  limma_fc = 2
)
```

# Pathway Tab Stuff
```{r}
gene_sets <- read_gene_sets(
  converted = converted,
  all_gene_names = gene_names,
  go = "GOBP",
  select_org = "BestMatch",
  idep_data = idep_data,
  my_range = c(15, 2000)
)

gage <- gage_data(
  select_go = "GOBP",
  select_contrast = "I:p53_null.Treatment_IR",
  min_set_size = 15,
  max_set_size = 2000,
  limma = limma,
  gene_p_val_cutoff = 1,
  gene_sets = gene_sets,
  absolute_fold = FALSE,
  pathway_p_val_cutoff = .1,
  n_pathway_show = 30
)

contrast_samples_pgsea <- find_contrast_samples(
  select_contrast = "I:p53_null.Treatment_IR", 
	all_sample_names = colnames(processed_data$data),
	sample_info = loaded_data$sample_info,
	select_factors_model = c("p53", "Treatment"),
	select_model_comprions = "p53: null vs. wt", 
  reference_levels = c("p53:wt", "Treatment:mock"),
	counts_deg_method = "3",
	data_file_format = "1"
)

#plot_pgsea(
  #my_range = c(15, 2000),
  #processed_data = processed_data$data,
  #contrast_samples = contrast_samples_pgsea,
  #gene_sets = gene_sets,
  #pathway_p_val_cutoff = .1,
  #n_pathway_show = 30
#)

fgsea <- fgsea_data(
  select_contrast = "I:p53_null.Treatment_IR",
  my_range = c(15, 2000),
  limma = limma,
  gene_p_val_cutoff = 1,
  gene_sets = gene_sets,
  absolute_fold = FALSE,
  pathway_p_val_cutoff = .1,
  n_pathway_show = 30
)

reactome_data(
  select_contrast = "I:p53_null.Treatment_IR",
  my_range = c(15, 2000),
  limma = limma,
  gene_p_val_cutoff = 1,
  converted = converted,
  idep_data = idep_data,
  pathway_p_val_cutoff = .1,
  n_pathway_show = 30,
  absolute_fold = FALSE
)

#pgsea_plot_all(
  #go = "GOBP",
  #my_range = c(15, 2000),
  #data = processed_data$data,
  #select_contrast = "I:p53_null.Treatment_IR",
  #gene_sets = gene_sets,
  #pathway_p_val_cutoff = .1,
  #n_pathway_show = 30
#)

pgsea_plot_data <- get_pgsea_plot_data(
  my_range = c(15, 2000),
  data = processed_data$data,
  select_contrast = "I:p53_null.Treatment_IR",
  gene_sets = gene_sets,
  sample_info = loaded_data$sample_info,
  select_factors_model = c("p53", "Treatment"),
  select_model_comprions = "p53: null vs. wt",
  pathway_p_val_cutoff = .1,
  n_pathway_show = 30
) 

gage[, 2]

select_data <- pathway_select_data(
  sig_pathways = "RNA processing ",
  gene_sets = gene_sets,
  contrast_samples = contrast_samples_pgsea,
  data = processed_data$data,
  select_org = "BestMatch",
  all_gene_names = gene_names
)

basic_heatmap(
  data = select_data,
  c("green", "black", "red")
)

pathway_list_data <- get_pathway_list_data(
  pathway_method = 1,
  gage_pathway_data = gage,
  fgsea_pathway_data = NULL,
  pgsea_plot_data = NULL,
  pgsea_plot_all_samples_data = NULL,
  go = "GOBP",
  select_org = "BestMatch",
  gene_info = all_gene_info,
  gene_sets = gene_sets
)

enrichment_plot(
  go_table = pathway_list_data
)

kegg_pathway(
  go = "KEGG",
  gage_pathway_data = gage,
  sig_pathways = "Oocyte meiosis",
  select_contrast = "I:p53_null.Treatment_IR",
  limma = limma,
  converted = converted,
  idep_data = idep_data,
  select_org = "BestMatch"
)
```

# Genome Tab
```{r}
all_gene_info <- get_gene_info(
  converted = converted,
  select_org = "BestMatch",
  gene_info_files = idep_data$gene_info_files
)

chromosome_plotly(
  limma = limma,
  select_contrast = "I:p53_null.Treatment_IR",
  all_gene_info = all_gene_info,
  ignore_non_coding = FALSE,
  limma_p_val_viz = .1,
  limma_fc_viz = 4,
  label_gene_symbol = FALSE,
  ma_window_size = 6,
  ma_window_steps = 2,
  ch_region_p_val = .01
)

genome_plot_data_pre <- get_genome_plot_data_pre(
  select_contrast = "I:p53_null.Treatment_IR",
  limma = limma,
  all_gene_info = all_gene_info
)

genome_plot_data <- get_genome_plot_data(
  genome_plot_data_pre = genome_plot_data_pre,
  all_gene_info = all_gene_info,
  select_contrast = "I:p53_null.Treatment_IR",
  limma = limma,
  regions_p_val_cutoff = .01,
  statistic_cutoff = .5
)

get_genome_plot(
  genome_plot_data = genome_plot_data,
  regions_p_val_cutoff = .01,
  statistic_cutoff = .5
)
```

# Bicluster Functions
```{r}
#"BCCC()" biclust
#"BCQU()" QUBIC
#"BCUnibic()" runibic
#"BCXmotifs()" biclust
#"BCPlaid()" biclust
#"BCSpectral()" biclust
#"BCBimax()" biclust
#"BCQuest()" biclust
biclustering <- get_biclustering(
  data = processed_data$data,
  n_genes = 1000,
  biclust_method = "biclust::BCCC()"
)

biclust_data <- biclust::bicluster(biclustering$data, biclustering$res, 1)[[1]]

basic_heatmap(
  data = biclust_data,
  c("green", "black", "red")
)

biclust_gene_names <- merge_data(
  all_gene_names = gene_names,
  data = biclust_data,
  merge_ID = "ensembl_ID"
)
# Only keep the gene names and scrap the data
gene_names_query <- dplyr::select_if(biclust_gene_names, is.character)

gene_sets <- read_pathway_sets(
  all_gene_names_query = gene_names_query,
  converted = converted,
  go = "GOBP",
  select_org = "BestMatch",
  gmt_file = NULL,
  idep_data = idep_data,
  gene_info = all_gene_info
)

pathway_info <- find_overlap(
  pathway_table = gene_sets$pathway_table,
  query_set = gene_sets$query_set,
  total_genes = gene_sets$total_genes,
  processed_data = processed_data$data,
  gene_info = all_gene_info,
  go = "GOBP",
  idep_data = idep_data,
  select_org = "GOBP",
  sub_pathway_files = gene_sets$pathway_files,
  use_filtered_background = FALSE,
  reduced = FALSE
)

biclust_table <- get_biclust_table_data(
  res = biclustering$res,
  biclust_data = biclust_data,
  select_go = "GOBP",
  select_org = "BestMatch",
  all_gene_info = all_gene_info
)
```

# NETWORK TAB
```{r}
wgcna <- get_wgcna(
  data = processed_data$data,
  n_genes = 1000,
  soft_power = 5,
  min_module_size = 20
)

get_wgcna_modules(wgcna)

network_plot <- get_network_plot(
  select_wgcna_module = "1. turquoise (495 genes)",
  wgcna = wgcna,
  top_genes_network = 10,
  select_go = "GOBP",
  select_org = "BestMatch",
  all_gene_info = all_gene_info,
  edge_threshold = .4
)

network_plot()

network_query <- network_enrich_data(
  select_wgcna_module = "1. turquoise (495 genes)",
  wgcna = wgcna
)

gene_query <- dplyr::filter(gene_names, ensembl_ID %in% network_query)

gene_sets <- read_pathway_sets(
  all_gene_names_query = gene_query,
  converted = converted,
  go = "GOBP",
  select_org = "BestMatch",
  gmt_file = NULL,
  idep_data = idep_data,
  gene_info = all_gene_info
)

pathway_info <- find_overlap(
  pathway_table = gene_sets$pathway_table,
  query_set = gene_sets$query_set,
  total_genes = gene_sets$total_genes,
  processed_data = processed_data$data,
  gene_info = all_gene_info,
  go = "GOBP",
  idep_data = idep_data,
  select_org = "BestMatch",
  sub_pathway_files = gene_sets$pathway_files,
  use_filtered_background = TRUE,
  reduced = FALSE
)

plot_scale_independence(
  wgcna = wgcna
)

plot_mean_connectivity(
  wgcna = wgcna
)
```

