---
title: "Pre_Process Workflow"
output: pdf_document
params:
  loaded_data:
  sample_info:
  data_file_format:
    label: "Format"
    value: 2
    input: radio
    choices: [1,2,3] 
  no_id_conversion:
    label: "Do not convert to ensembl"
    value: FALSE
    input: checkbox
  min_counts:
    label: "min PCM"
    value: .5
    input: numeric
  n_min_samples_count:
    label: "n libaries"
    value: 1
    input: numeric
  counts_transform:
    label: "Transform counts data for clustering & PCA."
    value: 1
    input: radio
    choices: [edgeR, vst, rlog]
  counts_log_start:
    label: "psuedocount"
    value: 4
    input: numeric
  log_transform_fpkm:
    label: "(Normalized Expression) Log transform?"
    value: "No"
    input: radio
    choiceNames: ["No", "Yes"]
    choiceValues: [FALSE, TRUE]
  log_start_fpkm:
    label: "(Normalized Expression) Constant c for started log(x+c)"
    value: 3
    input: numeric
  low_filter_fpkm:
    label: "(Normalized Expression) fpkm Minimum level"
    value: .1
    input: numeric
  n_min_samples_fpkm:
    label: "(Normalized Expression) Minimum Samples"
    value: 1
    input: numeric
  missing_value:
    label: "missing value imputation"
    value: geneMedian
    input: select
    choices: ["geneMedian", "treatAsZero", "geneMedianInGroup"]
  scatter_x:
    label: "X axis"
    value: ""
    input: text
  scatter_y:
    label: "Y axis"
    value: ""
    input: text
  sd_color:
    label: "SD_vs_mean"
    value: "Red"
    input: text
  rank:
    label: "Use mean of Rank values"
    value: False
    input: checkbox
  no_fdr:
    label: "[LFC] No P values"
    value: FALSE
    input: checkbox
  date:
    label: "Date: "
    value: !r Sys.Date()
  printcode:
    label: "Display Code"
    value: TRUE
    input: checkbox
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = params$printcode)
```
This pdf documents what parameters values were selected on the IDEP interface. It also includes the plots generated from those selections. 

<!-- This document cannot be knit directly from R studio -->
# Print All Selections
```{r}
for (i in 1:length(params)){
  if(names(params)[i] != "sample_info" && names(params)[i] != "loaded_data") # exclude loaded data & sample info
    cat(paste0(names(params)[i], ": ", params[[i]], "\n"))
}


```
<!-- "Read counts data (recommended)" = 1 -->

<!-- "Normalized expression values (RNA-seq FPKM, microarray, etc.)" = 2, -->

<!-- "Fold-changes and corrected P values from CuffDiff or any other program" = 3 -->


<!-- counts_transform guide: -->

<!-- "VST: variance stabilizing transform" = 2 -->

<!-- "rlog: regularized log (slow) " = 3 -->

<!-- "EdgeR: log2(CPM+c)" = 1 -->


```{r, include = FALSE}
#devtools::load_all()
library(idepGolem)

#The data was`r 
# switch(params$data_file_format, "1" = paste0("transformed with", switch(params$counts_transform, "1" = paste0("EdgeR using a psudocount of ", params$counts_log_start), "2" = "VST: Variance Stabilizing Transformation", "3" = "Regularized log"), " and imputed missing values using ", params$missing_value, "."), "2" = paste0(switch(params$log_transform_fpkm, "No" = "not log transformed", "Yes" = paste0(" log transformed with a psuedocount of ", params$log_start_fpkm)), " then filtered to only keep genes above level ", params$low_filter_fpkm, " in at least ", params$n_min_samples_fpkm, "sample(s) and imputed missing values using ", params$missing_value, "."))

```

`r switch(params$data_file_format,
"1" = "Read Counts",
"2" = "Normalized data",
"3" = "LFC and P values")` were analyzed using iDEP v0.97 [Citation]. Data was first filtered to remove reads below `r params$min_counts` CPM in at least `r params$n_min_samples_count` sample(s). 

The data was`r switch(params$data_file_format, 
"1" = paste0("transformed with ", 
switch(params$counts_transform,
"1" = paste0("EdgeR using a psudocount of ", params$counts_log_start),
"2" = "VST: Variance Stabilizing Transformation", 
"3" = "Regularized log"), " and imputed missing values using ", params$missing_value, "."),
"2" = paste0(switch(params$log_transform_fpkm, "No" = "not log transformed", "Yes" = paste0(" log transformed with a psuedocount of ", params$log_start_fpkm)), " then filtered to only keep genes above level ", params$low_filter_fpkm, " in at least ", params$n_min_samples_fpkm, "sample(s) and imputed missing values using ", params$missing_value, "."))`

# Plots

```{r Pre-Process Data, include = FALSE }
processed_data <- idepGolem::pre_process(
  data = params$loaded_data,
  missing_value = params$missing_value,
  data_file_format = params$data_file_format,
  low_filter_fpkm = params$low_filter_fpkm,
  n_min_samples_fpkm = params$n_min_samples_fpkm,
  log_transform_fpkm = params$log_transform_fpkm,
  log_start_fpkm = params$log_start_fpkm,
  min_counts = params$min_counts,
  n_min_samples_count = params$n_min_samples_count,
  counts_transform = params$counts_transform,
  counts_log_start = params$counts_log_start,
  no_fdr = params$no_fdr
)
```


```{r }
if (params$data_file_format == 1){
  idepGolem::total_counts_ggplot(
    counts_data = processed_data$raw_counts,
    sample_info = params$sample_info
  )  
}

```


```{r fig.height=5.5, fig.width=7.5 }
idepGolem::eda_scatter(
  processed_data = processed_data$data,
  plot_xaxis = params$scatter_x,
  plot_yaxis = params$scatter_y
  
)
```


```{r fig.height=5.5, fig.width=7.5}
idepGolem::eda_boxplot(
  processed_data = processed_data$data,
  sample_info = params$sample_info
)
```

```{r }
idepGolem::eda_density(
  processed_data = processed_data$data,
  sample_info = params$sample_info
)
```

```{r}
idepGolem::mean_sd_plot(
  processed_data = processed_data$data,
  heat_cols = params$sd_color,
  rank = params$rank
)
```

