<h3>
    iDEP 2.0: interactive & reproducible
</h3>
<ul>
    <li>
        Exploratory analysis, D.E.G., and pathway analyses on bulk RNA-seq data, All you need is a gene count matrix.
     </li>
     <li>
       Rich, interactive visualization: clustering, KEGG pathways, genome view, etc.
    </li>
    <li>
     Download HTML reports for some steps, documenting parameters and results,
    </li>
    <li>
      Download dozens of publication-ready figures in minutes,
    </li>
 	<li>
        Download R code and data objects, which can be used in RStudio to reproduce the analysis,
    </li>
    <li>
      Install locally as an R package or as a Docker container 
	  (<a href="https://youtu.be/MEVh3PJ9dxk">Windows</a>, <a href="https://youtu.be/u8Gdog4VAGc">MacOS</a>)
    </li>
    <li>
        Upload a pathway file (GMT file) to perform pathway analysis using your own gene-sets or analyze data from unsupported species.
      </li>

     
     
</ul>

<h4>
  Old versions are still usable (<a href="http://bioinformatics.sdstate.edu/idep96">v0.96</a>, 
  <a href="http://bioinformatics.sdstate.edu/idep11">v1.13</a>).
  If this server is busy, try a mirror on JetStream2: <a href="http://Ge-lab.org/idep">Ge-lab.org/idep</a>.
</h4>
<p>
    Email    <a href="mailto:gelabinfo@gmail.com?Subject=iDEP"> Jenny (gelabinfo@gmail)</a> or report
issues on our    <a href="https://github.com/espors/idepGolem/issues">GitHub page.</a> iDEP is still under active development and testing. We welcome any
suggestions or questions. Supported by NHGRI (National Human Genome Research Institute), NIH.
</p>
<p>
    ABSOLUTELY NO WARRANTY. Please double check your results using other tools.
</p>
<br/>
